Starting the monty section (#6)
mrc-ide/odin-monty: monty.qmd

# Getting started with monty

```{r}
#| include: false
source("common.R")
```

The `monty` R package is designed for modular work with statistical distributions, leveraging Monte Carlo methods for sampling. It enables the construction of increasingly complex models through four approaches:

1. Simple R code,
2. A dedicated domain-specific language (DSL),
3. Integration of `odin` models,
4. Composition of two `monty` models.

The most basic method to define a `monty` model is through a simple R function, as shown in this chapter. More advanced examples are covered in the @sec-monty-models in particular in the context of Bayesian statistics. @sec-monty-dsl introduces the monty DSL for more versatile model building using a small probabilistic DSL similar to BUGs. @sec-monty-combine explains how to compose new monty models by combining two existing ones. The last part of this book starting from @sec-inference demonstrates how to incorporate `odin` models into `monty` workflows.

## A simple example

```{r}
library(monty)
```

> We can define a simple Gaussian [mixture model](https://en.wikipedia.org/wiki/Mixture_model) of two sub-populations using the `monty_model_function()` from the package. The model represents the distribution of a quantity $l$, sampled with a probability $p$ from a normal distribution of mean $m_{1}$ and with a probability $1-p$ from another normal distribution of mean $m_{2}$. Both subpopulations have variance equal to 1. 

Reviewer: I am sure we could come up with a less abstract example to help people think about this, but this can be done in the next PR

Author: I think there might be a nice biology example with antibody titres and positive vs negative subpopulations - but yes can wait for another iteration.

To build our `monty` model, we start by defining an R function returning the log-density of our statistical model. This function has for arguments $l$, the quantity of interest and the three parameters $p$, $m_{1}$ and $m_{1}$ defining the two Normally-distributed subpopulations.

```{r}
fn <- function(l, p, m1, m2) {

>   log(p*dnorm(l,mean = m1) + (1-p)*dnorm(l, mean = m2))

Reviewer: ```suggestion
  log(p * dnorm(l, mean = m1) + (1 - p) * dnorm(l, mean = m2))
```

Or usual style. I'll try and get a linter up on this soon

>   log(p*dnorm(l,mean = m1) + (1-p)*dnorm(l, mean = m2))

Reviewer 2: ```suggestion
  log(p * dnorm(l, mean = m1) + (1 - p) * dnorm(l, mean = m2))
```

}
```

Assuming that the population is 75% from subpopulation 1 (normally distributed with mean 3) and 25% from subpopulation 2 (also normally distributed but with mean 7), we can build our `monty` model by indicating that the parameters of the subpopulations are fixed at these given values.

```{r}

> l_distribution <- monty_model_function(fn,

Reviewer: Why `l_distribution` - can you use a better name here?

Author: Changed to "mixture_model"

>                                        fixed = list(p = 0.75, m1 = 3, m2=7))

Reviewer: ```suggestion
                                       fixed = list(p = 0.75, m1 = 3, m2 = 7),
                                       allow_multiple_parameters = TRUE)
```

this is what you need to void vectrisation (plus some consistency in spacing)

```

We have just created a `monty` model. 

```{r}
l_distribution
```

We can plot the density of our model for values of $l$ between 0 and 10 and eye-check that we can see the two subpopulations in the correct proportions and with the 3 and 7 as modes.

```{r}
l <- seq(from = 0, to = 10, by = 0.1)
plot(l,

>      exp(Vectorize(l_distribution$density)(l)),

Author: @richfitz I need to use Vectorize() here to pass multiple value of the argument - let me know if a better interface can be shown.

Reviewer: See above. You will need to pass a 1-row matrix in to vectorise , so

```r
l <- matrix(seq(from = 0, to = 10, by = 0.1), 1)
monty_distribution_density(l_distribution, l)
```

Please do not use `$density` in docs

Author: Not sure how to make this work. The following code:
`fn <- function(l, p, m1, m2) {
  log(p * dnorm(l, mean = m1) + (1 - p) * dnorm(l, mean = m2))
}
l_distribution <- monty_model_function(fn,
                                       fixed = list(p = 0.75, m1 = 3, m2 = 7),
                                       allow_multiple_parameters = TRUE)
l <- matrix(seq(from = 0, to = 10, by = 0.1), 1)
monty_model_density(l_distribution, l)`

Return the following error, that I interpret as a conflict with having some parameters set as fixed. I could pass a 101x4 instead with the fixed argument like this (that works fine),

`l_distribution <- monty_model_function(fn,
                                       allow_multiple_parameters = TRUE)

l <- matrix(c(seq(from = 0, to = 10, by = 0.1), rep(0.75,101), rep(3,101), rep(7,101)),4)
monty_model_density(l_distribution, l)`

but then it makes things quite more complex to understand and it's not the same as the monty model as then 4 "parameters" with a mixed interpretation instead of one.

`Error in `packer$unpack()`:
! Can't unpack a matrix where the unpacker uses 'fixed'
Run `rlang::last_trace()` to see where the error occurred.`

We want now to sample from this model, using the `monty_sample()` function. For this we need to tell `monty` which sampler we want to use to explore our distribution. There are a variety of samplers avalaible and you can learn about them in @sec-samplers. One of the simplest is the random walk [Metropolis-Hastings](https://en.wikipedia.org/wiki/Metropolis%E2%80%93Hastings_algorithm) algorithm that should work almost out of the box (though not necesseraly efficiently) in most cases.

The random walk sampler uses a variance-covariance (VCV) matrix to guide its exploration, determining the 'jump' from the current point to the next in a random walk by drawing from a multivariate normal distribution parameterised by this matrix. For a single-parameter model, we use a 1x1 matrix (`matrix(1)`) as our VCV matrix.

The choice of the VCV matrix is critical for the sampler's efficiency, especially in more complex cases where the tuning of this matrix can significantly affect performance. A well-chosen VCV matrix optimises moving accross the parameter space, making the random walk sampler more effective in exploring the distribution of interest.

```{r}

> sampler <- monty_sampler_random_walk(matrix(1))

Reviewer: People may get confused with this being `matrix(1)` (it's a 1x1 matrix with a value of 1). If you used `matrix(0.5)` (say) then it would be easier to talk about (a 1x1 matrix with a variance in the proposal of 0.5)

Author: Done. Changed to a variance of 2 - not much of a difference in sampling performance, with proposal steps on average sqrt(2) bigger.

```

> We are now ready to sample from our distribution using `monty_sample()` with 2000 samples, starting our MCMC chain at the value 3 and running 4 chains simultaneously.

Reviewer: ```suggestion
We are now ready to sample from our distribution using `monty_sample()` with 2000 samples, starting our MCMC chain at the value 3 and running 4 chains sequentially.
```

These run **sequentially**, not simultaneously

> We are now ready to sample from our distribution using `monty_sample()` with 2000 samples, starting our MCMC chain at the value 3 and running 4 chains simultaneously.

Reviewer 2: ```suggestion
We are now ready to sample from our distribution using `monty_sample()` with 2000 samples, starting our MCMC chain at the value 3 and running 4 chains sequentially.
```

```{r}
samples <- monty_sample(l_distribution, sampler, 2000, initial = 3, n_chains = 4)
```

We can visualise our 4 chains.
```{r}
matplot(samples$density, type = "l", lty = 1,
        xlab = "sample", ylab = "log posterior density", col = "#00000055")
```

We can also check that our samples are correctly representing our distribution:
```{r}

> hist(samples$pars["l",,], breaks=100)

Reviewer 2: ```suggestion
hist(samples$pars["l", , ], breaks = 100)
```

> ```{r}
> hist(samples$pars["l",,], breaks=100)
> ```

Reviewer: ````suggestion

```{r}
hist(samples$pars["l", , ], breaks = 100)
```
````

Author: Changed.

## Connection of `monty` with other software

> `monty` includes a simple probabilistic domain-specific language (DSL) that is inspired by languages of the BUGS family such as `stan` and [Statistical Rethinking](https://xcelab.net/rm/). It is designed to make some tasks a bit easier, particularly when defining priors for your model.  We expect that this DSL is not sufficiently advanced to represent most interesting models but it may get more clever and flexible in the future. In particular we do not expect the DSL to be useful in writing likelihood functions for comparison to data; we expect that if your model is simple enough for this you would be better off using `stan` or some similarly flexible system.

Reviewer: This section feels like it is now in the wrong place

Before showing how to fit odin models to data, we'll focus for a bit on monty itself.  If you are anxious to get on and fit the model from @sec-data-filter, you might skip ahead to @sec-inference, where we resume this.

The `monty` R package is designed for modular work with statistical distributions, leveraging Monte Carlo methods for sampling. It enables the construction of increasingly complex models through four approaches:

1. Simple R code,
2. A dedicated domain-specific language (DSL),
3. Integration of `odin` models,
4. Composition of two `monty` models.

> The most basic method to define a `monty` model is through a simple R function, as shown in this chapter. More advanced examples are covered in the @sec-monty-models in particular in the context of Bayesian statistics. @sec-monty-dsl introduces the monty DSL for more versatile model building using a small probabilistic DSL similar to BUGs. @sec-monty-combine explains how to compose new monty models by combining two existing ones. The last part of this book starting from @sec-inference demonstrates how to incorporate `odin` models into `monty` workflows.

Reviewer 2: Referring to sec-monty-combine here, doesn't seem to exist currently

Author: Removed reference.

## A simple example

```{r}
library(monty)
```

We can define a simple Gaussian [mixture model](https://en.wikipedia.org/wiki/Mixture_model) of two sub-populations using the `monty_model_function()` from the package. The model represents the distribution of a quantity $l$, sampled with a probability $p$ from a normal distribution of mean $m_{1}$ and with a probability $1-p$ from another normal distribution of mean $m_{2}$. Both subpopulations have variance equal to 1. 

> To build our `monty` model, we start by defining an R function returning the log-density of our statistical model. This function has for arguments $l$, the quantity of interest and the three parameters $p$, $m_{1}$ and $m_{1}$ defining the two Normally-distributed subpopulations.

Reviewer 2: ```suggestion
To build our `monty` model, we start by defining an R function returning the log-density of our statistical model. This function has four arguments $l$ (the quantity of interest) and the three parameters $p$, $m_{1}$ and $m_{2}$ defining the two Normally-distributed subpopulations.
```

```{r}
fn <- function(l, p, m1, m2) {
  log(p * dnorm(l, mean = m1) + (1 - p) * dnorm(l, mean = m2))
}
```

Assuming that the population is 75% from subpopulation 1 (normally distributed with mean 3) and 25% from subpopulation 2 (also normally distributed but with mean 7), we can build our `monty` model by indicating that the parameters of the subpopulations are fixed at these given values.

```{r}
mixture_model <- monty_model_function(fn,
                                       fixed = list(p = 0.75, m1 = 3, m2 = 7),
                                       allow_multiple_parameters = TRUE)
```

We have just created a `monty` model. 

```{r}
mixture_model
```

> We can plot the density of our model for values of $l$ between 0 and 10 and eye-check that we can see the two subpopulations in the correct proportions and with the 3 and 7 as modes.

Reviewer 2: ```suggestion
We can plot the density of our model for values of $l$ between 0 and 10 and eye-check that we can see the two subpopulations in the correct proportions and with 3 and 7 as modes.
```

```{r}
#l <- matrix(seq(from = 0, to = 10, by = 0.1), 1)
l <- seq(from = 0, to = 10, by = 0.1)
#monty_model_density(l_distribution, l)
plot(l,
     exp(Vectorize(mixture_model$density)(l)),
     type = "l",
     ylab = "density")
```

## Sampling from our example distribution

> We now want to sample from this model, using the `monty_sample()` function. For this we need to tell `monty` which sampler we want to use to explore our distribution. There are a variety of samplers avalaible and you can learn about them in @sec-samplers. One of the simplest is the random walk [Metropolis-Hastings](https://en.wikipedia.org/wiki/Metropolis%E2%80%93Hastings_algorithm) algorithm that should work almost out of the box (though not necesseraly efficiently) in most cases.

Reviewer 2: ```suggestion
We now want to sample from this model, using the `monty_sample()` function. For this we need to tell `monty` which sampler we want to use to explore our distribution. There are a variety of samplers avaliable and you can learn about them in @sec-samplers. One of the simplest is the random walk [Metropolis-Hastings](https://en.wikipedia.org/wiki/Metropolis%E2%80%93Hastings_algorithm) algorithm that should work almost out of the box (though not necesseraly efficiently) in most cases.
```

Author: Good spot - but not still not the right spelling due to malignant typo

Author: Corrected though ;)

```

We can also check that our samples are correctly representing our distribution:
```{r}
hist(samples$pars["l", , ], breaks = 100)
```

## Connection of `monty` with other software

`monty` includes a simple probabilistic domain-specific language (DSL) that is inspired by languages of the BUGS family such as `stan` and [Statistical Rethinking](https://xcelab.net/rm/). It is designed to make some tasks a bit easier, particularly when defining priors for your model.  We expect that this DSL is not sufficiently advanced to represent most interesting models but it may get more clever and flexible in the future. In particular we do not expect the DSL to be useful in writing likelihood functions for comparison to data; we expect that if your model is simple enough for this you would be better off using `stan` or some similarly flexible system.

*mention SSM, dr jacoby, mcstate, BayesTools

## Going further

We've discussed a little bit about the philosophy of `monty` and built a simple model using an R function with `monty_function`. The example introduces the basics of defining and sampling from a `monty` model.

> For more advanced applications, refer to @sec-monty-dsl for constructing models using `monty`'s domain-specific language, which enables greater flexibility for complex structures. To combine multiple models, see @sec-monty-combine, which demonstrates how to compose and integrate existing `monty` models. Lastly, @sec-inference illustrates how to incorporate `odin` models into `monty` workflows, expanding the package’s potential for Bayesian inference with more sophisticated, system-based models. Each section provides detailed examples to guide you in leveraging `monty`’s full capabilities.

Reviewer 2: Again, referring to sec-monty-combine here when it doesn't seem to exist

Author: Removed ref.
